File: calendarview/src/main/java/al/bruno/calendar/view/adapter/CustomAdapter.java
package al.bruno.calendar.view.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;

public class CustomAdapter<T, VM extends ViewDataBinding> extends RecyclerView.Adapter<CustomViewHolder<T, VM>> {
    private List<T> t;
    private List<T> o;
    private int resources;
    private BindingInterface<T, VM> bindingInterface;

    public CustomAdapter(List<T> t, int resources, BindingInterface<T, VM> bindingInterface) {
        this.t = t;
        this.o = t;
        this.resources = resources;
        this.bindingInterface = bindingInterface;
    }
    @NonNull
    @Override
    public CustomViewHolder<T, VM> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CustomViewHolder<>(LayoutInflater.from(parent.getContext()).inflate(resources, parent, false), bindingInterface);

    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder<T, VM> holder, int position) {
        holder.bindData(t.get(position));
    }

    @Override
    public int getItemCount() {
        return t == null ? 0 : t.size();
    }

    public List<T> getItems() {
        return t;
    }

    public T getItem(int position) {
        return o.get(o.indexOf(t.get(position)));
    }

    public T getItem(T t) {
        return o.get(o.indexOf(t));
    }

    public int index(int position)
    {
        return o.indexOf(t.get(position));
    }
    public int index(T t1) {
        return o.indexOf(t1);
    }

    public void remove(T t1) {
        t.remove(t1);
        notifyItemRemoved(index(t1));
    }

    public void remove(int position) {
        t.remove(position);
        notifyItemRemoved(position);
    }

    public void notifyItemChanged(T t1) {
        notifyItemChanged(index(t1), t1);
    }

    public void update(T t1) {
        t.set(index(t1), t1);
        notifyItemChanged(index(t1), t1);
    }

    public void add(T t1){
        t.add(t1);
        //o.add(t1);
        notifyItemInserted(index(t1));
    }

    public void addAll(List<T> t1){
        t = t1;
        o = t1;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/al/bruno/financaime/adapter/CustomAdapter.kt
package al.bruno.financaime.adapter

import al.bruno.financaime.callback.BindingData
import al.bruno.financaime.holder.CustomViewHolder
import al.bruno.financaime.observer.Observer
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

class CustomAdapter<T, VM: ViewDataBinding>(t:List<T>, r:Int, bindingData:BindingData<T, VM>): RecyclerView.Adapter<CustomViewHolder<T, VM>>(), Observer<T> {

    private val t:ArrayList<T>;
    private val o:List<T>;
    private val r:Int
    private val bindingData:BindingData<T, VM>
    private var filter: Filter? = null;
    init {
        this.t = ArrayList(t)
        this.o = t
        this.r = r;
        this.bindingData = bindingData;
    }
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder<T, VM> {
        return CustomViewHolder(LayoutInflater.from(parent.context).inflate(r, parent, false), bindingData)
    }

    override fun getItemCount(): Int {
        return t.size;
    }

    override fun onBindViewHolder(viewHolder: CustomViewHolder<T, VM>, position: Int) {
        viewHolder.bind(t.get(position))
    }

    fun item(position: Int) : T {
        return t[position]
    }

    override fun update(l: T) {
        notifyItemChanged(t.indexOf(l))
    }

    override fun remove(l: T) {
        // because item was removed and mismatch the index
        // val position:Int = t.indexOf(l)
        t.remove(l)
        notifyItemRemoved(o.indexOf(l))
    }

    override fun add(l: T) {
        val position:Int = t.indexOf(l)
        when (position) {
            -1 -> {
                t.add(l)
                notifyDataSetChanged()
            }
            else -> {
                t[position] = l
                notifyItemChanged(position)
            }
        }
    }
}